apps/frontend/src/libs/components/input/input.tsx

```diff
@@ -5,35 +5,79 @@ import {
 	type FieldValues,
 } from "react-hook-form";
 
+import { getValidClassNames } from "~/libs/helpers/helpers.js";
 import { useFormController } from "~/libs/hooks/hooks.js";
 
+import styles from "./styles.module.css";
+
 type Properties<T extends FieldValues> = {
 	control: Control<T, null>;
 	errors: FieldErrors<T>;
 	label: string;
+	leftIcon?: JSX.Element;
 	name: FieldPath<T>;
 	placeholder?: string;
-	type?: "email" | "text";
+	rightIcon?: JSX.Element;
+	type?: "email" | "password" | "text";
 };
 
 const Input = <T extends FieldValues>({
 	control,
 	errors,
 	label,
+	leftIcon,
 	name,
 	placeholder = "",
+	rightIcon,
 	type = "text",
 }: Properties<T>): JSX.Element => {
 	const { field } = useFormController({ control, name });
-
 	const error = errors[name]?.message;
 	const hasError = Boolean(error);
 
+	const inputClassNames = getValidClassNames(
+		styles["input-field"],
+		leftIcon && styles["with-left-icon"],
+		rightIcon && styles["with-right-icon"],
+	);
+
 	return (
-		<label>
-			<span>{label}</span>
-			<input {...field} placeholder={placeholder} type={type} />
-			{hasError && <span>{error as string}</span>}
+		<label className={styles["input-label"]}>
+			<span className={styles["input-label-text"]}>{label}</span>
```

Comment:
I think we don't need span here, we can use only label

Reply:
Do you mean like this?
```
<label className={styles["input-label"]}>
	{label}
	...
</label>
```
Is it a better practice? I understand that less html is generally better, but for styling purposes - having an extra element can be more more appealing when debugging. Alternatively, is it better to pass styles to the container (label) and overwrite them? I don't think the difference in performance would be noticeable.

Reply:
thank you for this comprehensive answer, really appreciate this 🐾